notebooks/03_evaluate/comparison_movielens.ipynb

```diff
@@ -0,0 +1,982 @@
+    "\n",
+    "def train_sar_single_node(params, data):\n",
+    "    model = SARSingleNode(**params)\n",
+    "    model.set_index(data)    \n",
+    "    with Timer() as t:\n",
+    "        model.fit(data)\n",
+    "    return model, t\n",
+    "    \n",
+    "def train_ncf(params, data):\n",
+    "    model = NCF(n_users=data.n_users, n_items=data.n_items, **params)\n",
+    "    with Timer() as t:\n",
+    "        model.fit(data)\n",
+    "    return model, t\n",
+    "\n",
+    "    \n",
+    "trainer = {\n",
+    "    \"als\": lambda params, data: train_als(params, data),\n",
+    "    \"svd\": lambda params, data: train_svd(params, data),\n",
+    "    \"sar_single_node\": lambda params, data: train_sar_single_node(params, data), \n",
+    "    \"fastai\": lambda params, data: train_fastai(params, data),\n",
+    "    \"ncf\": lambda params, data: train_ncf(params, data),\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Regression predictions"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def predict_als(model, test):\n",
+    "    with Timer() as t:\n",
+    "        preds = model.transform(test)\n",
+    "    return preds, t\n",
+    "\n",
+    "\n",
+    "def predict_svd(model, test):\n",
+    "    with Timer() as t:\n",
+    "        preds = [model.predict(row[USER_COL], row[ITEM_COL], row[RATING_COL])\n",
+    "                       for (_, row) in test.iterrows()]\n",
+    "        preds = pd.DataFrame(preds)\n",
+    "        preds = preds.rename(index=str, columns={\"uid\": USER_COL, \n",
+    "                                                 \"iid\": ITEM_COL,\n",
+    "                                                 \"est\": PREDICTION_COL})\n",
+    "        preds = preds.drop([\"details\", \"r_ui\"], axis=\"columns\")\n",
+    "        for col in [USER_COL, ITEM_COL]:\n",
+    "            preds[col] = preds[col].astype(int)\n",
+    "    return preds, t\n",
+    "\n",
+    "    \n",
+    "def predict_fastai(model, test):\n",
+    "    with Timer() as t:\n",
+    "        preds = score(model, \n",
+    "                      test_df=test, \n",
+    "                      user_col=USER_COL, \n",
+    "                      item_col=ITEM_COL, \n",
+    "                      prediction_col=PREDICTION_COL)\n",
+    "    return preds, t\n",
+    "\n",
+    "\n",
+    "rating_predictor = {\n",
+    "    \"als\": lambda model, test: predict_als(model, test),\n",
+    "    \"svd\": lambda model, test: predict_svd(model, test),\n",
+    "    \"fastai\": lambda model, test: predict_fastai(model, test),\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Ranking predictions"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def recommend_k_als(model, test, train):\n",
```

Comment:
This is not needed because ALS has its own recommend_k function:
https://spark.apache.org/docs/2.2.0/api/python/pyspark.ml.html#pyspark.ml.recommendation.ALSModel.recommendForAllUsers